libcudacxx/include/cuda/std/detail/libcxx/include/variant

```diff
@@ -333,294 +429,315 @@ constexpr size_t __find_index() {
-    return __visit_alt(
-        __make_value_visitor(_CUDA_VSTD::forward<_Visitor>(__visitor)),
-        _CUDA_VSTD::forward<_Vs>(__vs)...);
+  template <class _Rp, class _Visitor, class... _Vs>
+  inline _LIBCUDACXX_HIDDEN _LIBCUDACXX_INLINE_VISIBILITY static constexpr _Rp
+  __visit_value(_Visitor&& __visitor, _Vs&&... __vs)
+  {
+    const size_t __first_index = __get_runtime_index<sizeof...(_Vs), 0>(__vs...);
+    return __visit_impl(
+      integer_sequence<size_t>{},
+      integer_sequence<size_t, (__remove_cvref_t<_Vs>::__size() - 1)...>{},
+      __first_index,
+      __make_value_visitor<_Rp>(_CUDA_VSTD::forward<_Visitor>(__visitor)),
+      _CUDA_VSTD::forward<_Vs>(__vs)...);
   }
 
 private:
   template <class _Visitor, class... _Values>
-  static constexpr void __std_visit_exhaustive_visitor_check() {
-    static_assert(is_invocable_v<_Visitor, _Values...>,
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr void __std_visit_exhaustive_visitor_check()
+  {
+    static_assert(_LIBCUDACXX_TRAIT(is_invocable, _Visitor, _Values...),
                   "`std::visit` requires the visitor to be exhaustive.");
   }
 
   template <class _Visitor>
-  struct __value_visitor {
+  struct __value_visitor
+  {
+    template <class... _Alts>
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr decltype(auto) operator()(_Alts&&... __alts) const
+    {
+      __std_visit_exhaustive_visitor_check<_Visitor, decltype((_CUDA_VSTD::forward<_Alts>(__alts).__value))...>();
+      return _CUDA_VSTD::__invoke(
+        _CUDA_VSTD::forward<_Visitor>(__visitor), _CUDA_VSTD::forward<_Alts>(__alts).__value...);
+    }
+    _Visitor&& __visitor;
+  };
+
+  template <class _Rp, class _Visitor>
+  struct __value_visitor_return_type
+  {
+    template <class... _Alts>
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr _Rp operator()(_Alts&&... __alts) const
+    {
+      __std_visit_exhaustive_visitor_check<_Visitor, decltype((_CUDA_VSTD::forward<_Alts>(__alts).__value))...>();
+      return _CUDA_VSTD::__invoke(
+        _CUDA_VSTD::forward<_Visitor>(__visitor), _CUDA_VSTD::forward<_Alts>(__alts).__value...);
+    }
+    _Visitor&& __visitor;
+  };
+
+  template <class _Visitor>
+  struct __value_visitor_return_type<void, _Visitor>
+  {
     template <class... _Alts>
-    inline _LIBCUDACXX_INLINE_VISIBILITY
-    constexpr decltype(auto) operator()(_Alts&&... __alts) const {
-      __std_visit_exhaustive_visitor_check<
-          _Visitor,
-          decltype((_CUDA_VSTD::forward<_Alts>(__alts).__value))...>();
-      return __invoke(_CUDA_VSTD::forward<_Visitor>(__visitor),
-                                _CUDA_VSTD::forward<_Alts>(__alts).__value...);
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr void operator()(_Alts&&... __alts) const
+    {
+      __std_visit_exhaustive_visitor_check<_Visitor, decltype((_CUDA_VSTD::forward<_Alts>(__alts).__value))...>();
+      _CUDA_VSTD::__invoke(_CUDA_VSTD::forward<_Visitor>(__visitor), _CUDA_VSTD::forward<_Alts>(__alts).__value...);
```

Comment:
We _also_ need to check that all the return types match. I wonder if this means that we should drop `__value_visitor`,  have the returntypeless version of `__make_value_visitor` compute the return type of a call to the visitor with an arbitrary choice of alternatives, make `__value_visitor_return_type` be returned from both flavors of `__make_value_visitor`, and have `__std_visit_exhaustive_visitor_check` also compare the type of the checked expression with `_Rp`?

(I know you're aware of the need, just suggesting a solution that seems the most ergonomic to me right now.)

Reply:
I believe this is done directly in `__visit_value` because we are effectively daisy-chaining all possible combinations into one function with a `decltype(auto)` return type deduction. Its essentially a big chain of `if (cond) return function` If the return types would not match the compiler would abort because it cannot deduce the return type of the whole function.

Its a bit unfortunate that it is implicit in the construction of the recursive algorithm but given the compile time implications I would rather not duplicate it

Reply:
:+1: added a comment